File: leetcode/0993-cousins-in-binary-tree/Solution.java
import java.util.ArrayDeque;
import java.util.Queue;

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public boolean isCousins(TreeNode root, int x, int y) {
        Queue<TreeNode> nodeQueue = new ArrayDeque<>();
        Queue<TreeNode> parentQueue = new ArrayDeque<>();
        nodeQueue.add(root);
        parentQueue.add(null);

        while (!nodeQueue.isEmpty()) {
            int levelSize = nodeQueue.size();
            TreeNode foundParentX = null;
            TreeNode foundParentY = null;
            boolean foundX = false;
            boolean foundY = false;

            for (int i = 0; i < levelSize; i++) {
                TreeNode node = nodeQueue.poll();
                TreeNode parent = parentQueue.poll();

                if (node.val == x) {
                    foundX = true;
                    foundParentX = parent;
                }
                if (node.val == y) {
                    foundY = true;
                    foundParentY = parent;
                }

                if (node.left != null) {
                    nodeQueue.add(node.left);
                    parentQueue.add(node);
                }
                if (node.right != null) {
                    nodeQueue.add(node.right);
                    parentQueue.add(node);
                }
            }

            if (foundX && foundY) {
                return foundParentX != foundParentY;
            }
            if (foundX || foundY) {
                return false;
            }
        }

        return false;
    }
}

File: leetcode/0993-cousins-in-binary-tree/solution.py
from typing import Optional
from collections import deque


# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def isCousins(self, root: Optional[TreeNode], x: int, y: int) -> bool:
        queue = deque([(root, None)])

        while queue:
            level_size = len(queue)
            found = []

            for _ in range(level_size):
                node, parent = queue.popleft()
                if node.val == x or node.val == y:
                    found.append(parent)
                if node.left:
                    queue.append((node.left, node))
                if node.right:
                    queue.append((node.right, node))

            if len(found) == 2:
                return found[0] is not found[1]
            if len(found) == 1:
                return False

        return False
